group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java

```diff
@@ -2377,156 +2606,2165 @@ public void testOnNewMetadataImage() {
-            ConsumerGroupCurrentMemberAssignmentValue expectedValue =
-                (ConsumerGroupCurrentMemberAssignmentValue) expected.message();
-            ConsumerGroupCurrentMemberAssignmentValue actualValue =
-                (ConsumerGroupCurrentMemberAssignmentValue) actual.message();
+        List<CompletableFuture<JoinGroupResponseData>> responseFutures = new ArrayList<>();
+        // First round of join requests. Generate member ids.
+        for (int i = 0; i < groupMaxSize + 1; i++) {
+            if (i == 0) {
+                responseFutures.add(context.sendGenericGroupJoin(
+                    request,
+                    requiredKnownMemberId,
+                    false,
+                    new ExpectedGenericGroupResult(Errors.NONE, true)
+                ));
+            } else {
+                responseFutures.add(context.sendGenericGroupJoin(request, requiredKnownMemberId));
+            }
+        }
 
-            assertEquals(expectedValue.memberEpoch(), actualValue.memberEpoch());
-            assertEquals(expectedValue.previousMemberEpoch(), actualValue.previousMemberEpoch());
-            assertEquals(expectedValue.targetMemberEpoch(), actualValue.targetMemberEpoch());
-            assertEquals(expectedValue.error(), actualValue.error());
-            assertEquals(expectedValue.metadataVersion(), actualValue.metadataVersion());
-            assertEquals(expectedValue.metadataBytes(), actualValue.metadataBytes());
+        GenericGroup group = context.groupMetadataManager.getOrMaybeCreateGenericGroup(
+            "group-id",
+            false
+        );
 
-            // We transform those to Maps before comparing them.
-            assertEquals(fromTopicPartitions(expectedValue.assignedPartitions()),
-                fromTopicPartitions(actualValue.assignedPartitions()));
-            assertEquals(fromTopicPartitions(expectedValue.partitionsPendingRevocation()),
-                fromTopicPartitions(actualValue.partitionsPendingRevocation()));
-            assertEquals(fromTopicPartitions(expectedValue.partitionsPendingAssignment()),
-                fromTopicPartitions(actualValue.partitionsPendingAssignment()));
-        } else {
-            assertEquals(expected.message(), actual.message());
+        assertEquals(0, group.size());
+        assertEquals(groupMaxSize + 1, group.numPendingJoinMembers());
+        assertTrue(group.isInState(EMPTY));
+
+        List<String> memberIds = verifyGenericGroupJoinResponses(responseFutures, 0, Errors.MEMBER_ID_REQUIRED);
+        assertEquals(groupMaxSize + 1, memberIds.size());
+
+        // Second round of join requests with the generated member ids.
+        // One of them will fail, reaching group max size.
+        responseFutures = new ArrayList<>();
+        for (String memberId : memberIds) {
+            responseFutures.add(context.sendGenericGroupJoin(request.setMemberId(memberId), requiredKnownMemberId));
         }
```

Comment:
for all places i use the old for each / for loops, there is an error
`Variable used in lambda expression should be final or effectively final`
because i reuse variables (mainly JoinGroupRequestData & responseFutures). i can use new variables instead, would that be better?